src/sage/schemes/hyperelliptic_curves_smooth_model/hyperelliptic_constructor.py

```diff
@@ -0,0 +1,279 @@
+    allowed as input::
+
+        sage: E = HyperellipticCurveSmoothModel(x^3+1)
+        sage: E.genus()
+        1
+        sage: HyperellipticCurveSmoothModel(x)
+        Traceback (most recent call last):
+        ...
+        ValueError: arguments f = x and h = 0 must define a curve of genus at least one.
+
+    The following polynomials define a singular curve and are
+    not allowed as input::
+
+        sage: C = HyperellipticCurveSmoothModel(x^6 + 2*x - 1, 2*x - 2)
+        Traceback (most recent call last):
+        ...
+        ValueError: singularity in the provided affine patch
+
+    The constructor accepts a distinguished point as an argument::
+
+        sage: C = HyperellipticCurveSmoothModel(x^5 + x + 2, distinguished_point=(1, -2))
+        sage: C.distinguished_point()
+        (1 : -2 : 1)
+
+    """
+
+    # ---------------------------
+    # Internal Helper functions
+    # ---------------------------
+
+    def __genus(f, h):
+        """
+        Helper function to compute the genus of a hyperelliptic curve
+        defined by `y^2 + h(x)y = f(x)`.
+        """
+        # Some classes still have issues with degrees returning `int`
+        # rather than Sage Integer types
+        df = Integer(f.degree())
+        dh_2 = 2 * Integer(h.degree())
+        if dh_2 < df:
+            return (df - 1) // 2
+        return (dh_2 - 1) // 2
+
+    def __check_no_affine_singularities(f, h):
+        """
+        Helper function which determines whether there are any
+        affine singularities in the curve `y^2 + h(x)y = f(x)`.
+        """
+        if f.base_ring().characteristic() == 2:
+            if h.is_zero():
+                return False
+            elif h.is_constant():
+                return True
+            return h.gcd(f.derivative() ** 2 - f * h.derivative() ** 2).is_one()
+
+        if h.is_zero():
+            return f.gcd(f.derivative()).is_one()
+
+        g1 = h**2 + 4 * f
+        g2 = 2 * f.derivative() + h * h.derivative()
+        return g1.gcd(g2).is_one()
+
+    def __defining_polynomial(f, h):
+        """
+        Compute the (homogenised weighted projective) defining polynomial of
+        the hyperelliptic curve.
+        """
+        X, Y, Z = PolynomialRing(f.base_ring(), names="X, Y, Z").gens()
+
+        # Some classes still have issues with degrees returning `int`
+        d = max(Integer(h.degree()), (Integer(f.degree()) + 1) // 2)
+        F = sum(f[i] * X**i * Z ** (2 * d - i) for i in range(2 * d + 1))
+
+        if h.is_zero():
+            G = Y**2 - F
+        else:
+            H = sum(h[i] * X**i * Z ** (d - i) for i in range(d + 1))
+            G = Y**2 + H * Y - F
+
+        return G
+
+    # -------------------------------------------
+    # Typechecking and projective model creation
+    # -------------------------------------------
+
+    # Check the polynomials are of the right type
+    F = h**2 + 4 * f
+    if not isinstance(F, Polynomial):
```

Comment:
Can we not annotate the input parameters with `Polynomial` as well? That doesn't mean you can remove this check, since type annotations are not checked at runtime, but it does let our CI catch any violations within library code, and it also makes things a bit more clear to users reading the documentation.